test/test_controller.py

```diff
@@ -1722,6 +1738,259 @@ def test_check_if_get_backup_list_fills_attributes_based_on_missing_files(
+    with mysql_cursor(**options) as cursor:
+        t0 = time.monotonic_ns()
+
+        cursor.execute(f"CREATE TABLE IF NOT EXISTS {table_name} (id INT NOT NULL PRIMARY KEY AUTO_INCREMENT, b LONGBLOB);")
+        cursor.execute("DROP PROCEDURE IF EXISTS generate_data;")
+        cursor.execute(
+            f"""
+        CREATE PROCEDURE generate_data()
+        BEGIN
+             DECLARE i INT DEFAULT 0;
+                WHILE i < {batches} DO
+                    INSERT INTO {table_name} (b) VALUES (REPEAT('x', {MB_PER_BATCH * ONE_MB}));
+                    SET i = i + 1;
+                END WHILE;
+        END
+        """
+        )
+        cursor.execute("CALL generate_data();")
+        cursor.execute("COMMIT")
+        cursor.execute("FLUSH BINARY LOGS")
+
+        t1 = time.monotonic_ns()
+
+        logger.info(
+            "Populating table %s with %i MB took %f sec.", table_name, (batches * MB_PER_BATCH), (t1 - t0) / 1_000_000_000
+        )
+
+
+def _get_table_size(mysql_config: MySQLConfig, table_name: str) -> int:
+    """Get size of table (data + index) in bytes."""
+    with mysql_cursor(**mysql_config.connect_options) as cursor:
+        cursor.execute(
+            f"""
+            SELECT TABLE_NAME AS `Table`,
+                (DATA_LENGTH + INDEX_LENGTH) AS `size`
+            FROM information_schema.TABLES
+            WHERE TABLE_NAME = '{table_name}';
+        """
+        )
+
+        return cast(SizeDict, cursor.fetchone())["size"]
+
+
+def do_backup(controller: Controller) -> list[BackupStream]:
+    """Trigger a backup and wait for it to finish."""
+    flow_tester = FlowTester(controller)
+
+    controller.switch_to_active_mode()
+    controller.start()
+
+    flow_tester.wait_for_streaming_binlogs()
+
+    # Stream backup.
+    controller.mark_backup_requested(backup_reason=BackupStream.BackupReason.requested)
+
+    flow_tester.wait_for_multiple_streams()
+    flow_tester.wait_for_streaming_binlogs()
+    flow_tester.wait_for_single_stream()
+
+    return controller.backup_streams
+
+
+def do_restore(
+    target_controller: Controller,
+    backup_streams: list[BackupStream],
+    caplog: LogCaptureFixture | None = None,
+    fail_because_disk_full: bool = False,
+) -> None:
+    """Trigger a restore and wait for it to finish."""
+    bs = backup_streams[0]
+
+    # Restore backup into an empty database.
+    flow_tester = FlowTester(target_controller, caplog=caplog)
+    target_controller.start()
+
+    try:
+        flow_tester.wait_for_fetched_backup(timeout=2)
+
+        target_controller.restore_backup(site=bs.site, stream_id=bs.stream_id)
+
+        if fail_because_disk_full:
+            flow_tester.wait_for_disk_full_being_logged()
+
+            # Check that we have backups, but none of them are broken.
+            current_backups = sort_completed_backups(target_controller.state["backups"])
+            assert current_backups
+            assert all(b["broken_at"] is None for b in current_backups)
```

Comment:
You should probably also check the state of the `restore_coordinator`, it should be `failed` and not `failed_basebackup` (which I assume it is if you run this check `MAX_BASEBACKUP_ERRORS` times).